---
title: "GAPIT example"
author: "Brian J. Knaus"
date: "`r format(Sys.time(), '%Y, %B %d')`"
output:
  html_document:
    toc: true
    code_folding: hide
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#knitr::opts_chunk$set(echo = FALSE)
knitr::opts_chunk$set(fig.align = "center")
knitr::opts_chunk$set(fig.height = 4)
knitr::opts_chunk$set(fig.width = 4)
st1 <- Sys.time()
```


## Functions


```{r}
#x<-"Phenotypes.txt"

read_GAPIT_phenotype <- function(x){
  myY <- read.table( file = x, header = TRUE, sep = "\t")
  if( !names(myY)[1] == "Taxa" ){
    stop( "First column header is not Taxa." )
  }
  myY$Taxa <- as.character(myY$Taxa)
  
  return(myY)
}

#x <- "Sacffold_3_2_hmp.txt"
read_GAPIT_hapmap <- function(x){
  myG <- read.table( file = x, header = FALSE, sep = "\t")
#  dput(as.vector(unlist(myG[1, 1:11])))
  eo_df <- data.frame(
    Expected = c("rs", "alleles", "chrom", "pos", "strand", "assembly", "center", 
"protLSID", "assayLSID", "panel", "QCcode"),
    Observed = as.vector(unlist(myG[1, 1:11]))
  )

  if( !all(eo_df$Expected == eo_df$Observed) ){
    print( eo_df )
    stop( "Unexpected columns." )
  }
  # myG[1:3, 1:4]
  return(myG)
}

# Y <- myY
# G <- myG

synchronize_GY <- function( G, Y, verbose = TRUE, sort = TRUE ){
  YnoG <- Y$Taxa[ !Y$Taxa %in% G[1, 12:ncol(G)] ]
  GnoY <- as.vector(unlist(G[1, 12:ncol(G)]))[ !G[1, 12:ncol(G)] %in% Y$Taxa ]
  
  Y <- Y[ Y$Taxa %in% G[1, 12:ncol(G)], , drop = FALSE ]
  Gtmp <- G[ , G[1, ] %in% Y$Taxa, drop = FALSE]
  G <- cbind( G[ , 1:11], Gtmp )
  
  if( sort ){
    Y$Taxa <- as.character(Y$Taxa)
    Y <- Y[sort.int(Y$Taxa, decreasing = FALSE, index.return = TRUE)$ix, ]
    
    Gtmp <- G[ , 12:ncol(G), drop = FALSE]
    Gtmp[1,] <- as.character(Gtmp[1,])
    Gtmp <- Gtmp[ , sort.int( as.vector(unlist(Gtmp[1, ])), decreasing = FALSE, index.return = TRUE)$ix]
    G <- cbind(G[, 1:11], Gtmp)
  }
  
#  Y$Taxa
  if( verbose ){
    if( length(YnoG) > 0 ){
      print("The following samples had phenotypes but no genotypes and have been omitted:")
      print(YnoG)
    } else {
      print("All phenotypes retained.")
    }
    if( length(GnoY) > 0 ){
      print("The following samples had genotypes but no phenotypes and have been omitted:")
      print(GnoY)
    } else {
      print("All genotypes retained.")
    }
  }
  
  return( list(G=G, Y=Y) )
}


```




## Phenotypes

```{r}
myY <- read_GAPIT_phenotype("Phenotypes.txt")
myY[1:3, ]
myY$Taxa[1:6]
```


## Genotypes


```{r}
myG <- read_GAPIT_hapmap("Scaffold_3_2_hmp.txt")

# Sanitize names
myG[1, 12:ncol(myG)] <- sub("-.+", "", myG[1, 12:ncol(myG)])

myG[1:3, c(1:4, 12:14)]
```


## Synchronize


```{r}

my_sync <- synchronize_GY( G = myG, Y = myY )

# Validate membership.
all(my_sync$G[1, 12:ncol(my_sync$G)] %in% my_sync$Y$Taxa)
# Validate membership and order.
all(my_sync$G[1, 12:ncol(my_sync$G)] == my_sync$Y$Taxa)

my_sync$Y$Taxa[1:10]
my_sync$G[1, c(12:21)]

```

## GAPIT


```{r}
library(GAPIT)

t1 <- Sys.time()

Analysis1 <- GAPIT(
  Y = my_sync$Y,
  G = my_sync$G,
  model = "MLM",
#  SNP.impute = "Major",
  SNP.impute = "Middle",
  PCA.total = 3,
#  Major.allele.zero=TRUE,
  Major.allele.zero=FALSE,
#  file.output = TRUE,
  file.output = FALSE,
)

t99 <- Sys.time()
t99 - t1

knitr::kable(Analysis1$GWAS[1:3, 1:5])

```




## Clean up


```{r}
# unlink("GAPIT.Association*")
# unlink("GAPIT.Genotype*")
# unlink("GAPIT.Phenotype*")
```

