---
title: "How to circlize"
author: "Brian J. Knaus"
date: "`r format(Sys.time(), '%Y, %B %d')`"
output:
  html_document:
    toc: true
    toc_depth: 2
    code_folding: show
editor_options: 
  chunk_output_type: console
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(fig.align = "center")
knitr::opts_chunk$set(fig.height = 4)
knitr::opts_chunk$set(fig.width = 4)
st1 <- Sys.time()
```


## Introduction


Circular plots have become a popular way to visualize genomic data.
This may be because the first genomes that were sequenced were bacterial genomes, which are circular.
Researchers of eukaryotic genomes have continued this style of graphic.
This is perhaps because it facilitates the comparison of chromosomes on the outside of the circle with bands that extend through the middle of the plot to indicate regions of synteny.
Early versions of this style of plot were created with the software 'Circos' (http://circos.ca/), in a system that draws on the organization present in Unix filesystems.
More recently, this plot style has been facilitated by the R package 'circlize' (https://CRAN.R-project.org/package=circlize).
This package is accompanied by a book (https://jokergoo.github.io/circlize_book/book/) to describe it's use.
Here we'll use the **Cannabis sativa** CBDRx assembly (https://www.ncbi.nlm.nih.gov/data-hub/genome/GCF_900626175.2/) as an example of creating circular plots in R.


## A simple plot


A basic plot can be created with 'BED' format data.
This format is described online (https://genome.ucsc.edu/FAQ/FAQformat.html#format1, https://uswest.ensembl.org/info/website/upload/bed.html, https://en.wikipedia.org/wiki/BED_(file_format)
).
This is tabular data where the first three columns indicate each chromosome name, start, and end coordinates.
Subsequent columns are optional.
We begin by summarizing the genomic FASTA sequence obtained from NCBI with a script I've created (https://github.com/knausb/fasta2nuccomp).
This results in a table of data that we can read in to R and convert to 'BED' format.


```{r}
nucs <- read.csv("./GCF_900626175.2_cs10_genomic_nuccomp.csv")
nucs <- nucs[1:10, ]

nucs$GC <- rowSums( nucs[, c("C", "c", "G", "g")] )/rowSums( nucs[, c("A", "a", "C", "c", "G", "g", "T", "t")] )
nucs$GC <- nucs$GC * 100


nucs[1:3, 1:8]
nrow(nucs)

myBed <- data.frame( chrom = nucs$Id, chromStart = 1, chromEnd = nucs$Length, gc = nucs$GC)

myBed$gcs <- myBed$gc - min(myBed$gc)
myBed$gcs <- myBed$gcs/max(myBed$gcs)
myBed$gcs <- myBed$gcs * 100
myBed$gcs <- as.integer(myBed$gcs)
myBed$gcs[ myBed$gcs == 0 ] <- 1
myBed$cols <- viridisLite::magma(n=100, alpha = 0.4, begin = 0.2, end = 1.0)[ myBed$gcs ]
#nSeq <- 10
#myBed <- data.frame( chrom = nucs$Id[1:nSeq], chromStart = 1, chromEnd = nucs$Length[1:nSeq])
# apply(myBed , MARGIN = 2, class)
head(myBed)
```

Genomic assemblies currently contain sequences that are close to full chromosome assemblies, as well as numerous shorter sequences we refer to as 'unplaced contigs.'
The hemp genome is expected to contain 10 chromosomes, so we've subset the data to these sequences.
Note that the BED format does not include a header line.
In R, it is frequently useful to use a header, so we've created one for our data.frame.


We now load the R package 'circlize'.
This package includes a verbose start up message.
The book shows us that we can suppress this message once we've read it.


```{r}
#library(circlize)
suppressPackageStartupMessages(library(circlize))
```


We can now create our plot.
Each plot begins with the function `circos.genomicInitialize()` and ends with `circos.clear()`.
The function `circos.genomicInitialize()` includes the BED-like data.frame we created and named 'myBed' as well as other parameters to control the appearance of the plot.
Note that 'circlize' uses the 'base' graphics engines as opposed to other R graphics engines (e.g., ggplot2 or lattice).
Parameters to help control the appearance of base R graphics can be found in the `?par` manual page.
Here we create our simple plot and add some text to it's center.


```{r, fig.width=6, fig.height=6, fig.cap="**Figure 1.** A simple circular plot."}
circos.genomicInitialize( 
  data = myBed,
  axis.labels.cex =  0.4 * par("cex"),
  labels.cex = 0.8 * par("cex") 
)

text(x = 0, y = 0.1, labels = "Industrial hemp: CBDRx")
text(x = 0, y = 0, labels = expression(paste(italic("Cannabis sativa"), " L.")))
text(x = 0, y = -0.12, labels = "(NCBI RefSeq: GCF_900626175.2)")

circos.clear()
```


## Tracks and sectors


A plot in 'circlize' may include 'tracks' and 'sectors'.
The **tracks** form concentric circles beginning at the outside of the circle and extend in towards the center of the plot.
In the context of our genomic plot, each track represents the genome (all 10 chromosomes).
Each track consists of **sectors** that each represent a plot or panel for each chromosome.
Within each panel we have an x and y coordinate system for plotting information.
We can also add `circos.par( start.degree = 90 )` to start thre plot at the top of the plot (12 o'clock).
We can now add to our first plot by adding two tracks that include 10 sectors.


```{r, fig.width=6, fig.height=6, fig.cap="**Figure 2.** A simple circular plot including tracks and sectors."}

circos.par( start.degree = 90 )
circos.genomicInitialize( 
  data = myBed,
  axis.labels.cex =  0.4 * par("cex"),
  labels.cex = 0.8 * par("cex") 
)

text(x = 0, y = 0.1, labels = "Industrial hemp: CBDRx")
text(x = 0, y = 0, labels = expression(paste(italic("Cannabis sativa"), " L.")))
text(x = 0, y = -0.12, labels = "(GCF_900626175.2)")

circos.track(ylim = c(0, 1), 
    bg.col = NA, 
    bg.border = "#00000044", 
    track.height = 0.1)

circos.track(ylim = c(0, 1), 
    bg.col = NA, 
    bg.border = "#00000044", 
    track.height = 0.15, bg.lty = 3, bg.lwd = 2)

circos.clear()
```


## GFF data


Genomic annotations (e.g., genes, transcripts, exons, etc.) are frequently provided in the GFF format (https://en.wikipedia.org/wiki/General_feature_format).
This can be seen as tabular data with a header region at the top (prefixed with # signs) and nine columns where the first eight are defined and the last column is intended to hold 'other' data in a 'key and value' format where multiple 'key and value' pairs are delimited by a semicolon (';').
For our purposes a GFF file can be read in with R's standard read functions, with a little modification.
A GFF file is tab delimited, we specify this by setting `sep = "\t"`.
The annotations may include content such as `5' UTR`.
The R read functions will tend to interpret the 'prime' as an unmatched single quote.
We can manage this by setting `quote = "\""` so only double quotes will be interpreted as quotes.

Our data includes chromosome scale sequences, as well as unplaced contigs and a mitochondrion.
The chromosomal sequences begin with "NC_04", we can use this to isolate the sequences we're interested in.


```{r}
gff <- read.table("GCF_900626175.2_cs10_genomic.gff.gz", sep = "\t", quote = "\"")
colnames(gff) <- c("seqid","source","type","start","end","score","strand","phase","attributes")
gff[1:3, 1:8]

# Isolate chromosomal sequences.
gff <- gff[grep("^NC_04", gff$seqid), ]

# Isolate 'genes.'
genes <- gff[grep("gene", gff$type), ]
nrow(genes)

# Isolate 'rRNA'.
rrna <- gff[grep("S rRNA", gff$attributes), ]
nrow(rrna)
```


The 9th column, or the attributes, of GFF data consists of 'key value' pairs delimited by a semicolon.
We can split this string on the delimiter (';') and use grep to return just the strings with the 'product' key.
We can then use this to isolate terpenes, or other products (keywords) we may be interested in.


```{r}
my_products <- lapply(strsplit(gff$attributes, split = ";"), function(x){ grep("product=", x, value = TRUE) })
my_products <- unlist(my_products)
my_products <- sub("product=", "", my_products)

#sort(table(my_products), decreasing = TRUE)[1:10]
#
sort(table(grep("[Tt]erpene", my_products, value = TRUE)), decreasing = TRUE)[1:6]

terps <- gff[grep("[Tt]erpene", gff$attributes), ]
nrow(terps)
```


## Gene abundance tracks


Now that we've gathered some information about our assembly, we can use it in our graphic.
We can use the gene start positions to create histograms.
The background for the histogram cells can be colored by GC content.
A track for rRNA and terpenes represented as points can demonstrate where these features are found.
Another track with rectangles may be useful for larger features.
Lastly, we can include labels for select features and links among features on different chromosomes.


```{r, fig.width=8, fig.height=8, fig.cap="**Figure 3.** A simple circular plot including tracks and sectors."}

#png(filename = "gdplot.png", width = 18, height = 18, units = "cm", res = 300)

# circos.clear()

circos.par( start.degree = 90 )

circos.genomicInitialize( 
  data = myBed,
  axis.labels.cex =  0.4 * par("cex"),
  labels.cex = 0.8 * par("cex") 
)

circos.trackHist( 
  sectors = genes$seqid,
#  sectors = as.factor(genes$seqid),
  x = genes$start,
  #border = "#FFFFFF",
  bg.border = "#FFFFFF",
  bg.col = myBed$cols,
  #col = c("#FF0000", "#00FF00"),
  #col = "#C0C0C0",
  col = "#808080",
  border = NA,
  breaks = seq(1, 1e9, by = 2e6) )

circos.track(ylim = c(0, 1),
    bg.col = NA,
    bg.border = "#00000044",
    track.height = 0.15, bg.lty = 3, bg.lwd = 2)

circos.trackPoints(
  sectors = as.factor(rrna$seqid),
  x = rrna$start,
  y = rep(0.5, times = nrow(rrna)),
  #y = genes$start,
  #pch = "|",
  pch = 18,
  col = "#B2222211"
)

circos.trackPoints(
  sectors = as.factor(terps$seqid),
  x = terps$start,
  y = rep(0.2, times = nrow(terps)),
  #y = genes$start,
  #pch = "|",
  pch = 17,
  cex = 1,
  col = "#1E90FF11"
)

circos.track(ylim = c(0, 1),
    bg.col = NA,
    bg.border = "#00000044",
    track.height = 0.10, bg.lty = 3, bg.lwd = 1)

for( i in unique(terps$seqid) ){
  tmp <- terps[terps$seqid == i, ]

  circos.rect(
    xleft = tmp$start,
    ybottom = -0.0,
    xright = tmp$end,
    ytop = 0.5,
    sector.index = i,
    col = "#1E90FF11",
    border = "#1E90FF11",
    lwd = 4
  )
}

circos.link("NC_044374.1", 0,
            "NC_044379.1", c(1e3,4e7),
            col = "#1E90FF66", 
            #border = "black", 
            lwd = 2, lty = 2)

circos.link("NC_044374.1", 0,
            "NC_044375.1", c(1e3,6e7),
            col = "#FFD70066", 
            #border = "black", 
            lwd = 2, lty = 2)

circos.labels(c("NC_044378.1", "NC_044378.1"),
              x = c(30.1e6, 28.12e6), labels = c("A", "CBDAS"),
    side = "inside")

myLeg <- rep("", times = 10)
myLeg[c(2, 9)] <- c("High GC%", "Low GC%")

legend('bottomleft',
       legend = myLeg,
       fill = viridisLite::magma(n=10, alpha = 0.4, begin = 1.0, end = 0.2),
       border = NA,
       pt.cex = 2,
       bty = "n", y.intersp = 0.5)

legend('bottomright',
       legend = c("rRNA", "Terpene"),
       col = c("#B22222", "#1E90FF"),
       border = NA,
       pch = c(18, 17),
       pt.cex = 2,
       bty = "n",
       y.intersp = 1.6)


title( main = "Industrial hemp: CBDRx\n(GCF_900626175.2)", line = -1 )

circos.clear()

#dev.off()

```


## Adding more...


```{r, fig.width=8, fig.height=8, fig.cap="**Figure 4.** A simple circular plot including tracks and sectors."}

#png(filename = "gdplot.png", width = 18, height = 18, units = "cm", res = 300)

# circos.clear()

circos.par( start.degree = 90 )
#circos.par(gap.after = c(2, 2, 2, 2, 10))
circos.par(gap.after = c(rep(2, times = 9), 10))

circos.genomicInitialize( 
  data = myBed,
  axis.labels.cex =  0.4 * par("cex"),
  labels.cex = 0.8 * par("cex") 
)

# text(x = 0, y = 0.1, labels = "Industrial hemp: CBDRx")
# text(x = 0, y = 0, labels = expression(paste(italic("Cannabis sativa"), " L.")))
# text(x = 0, y = -0.12, labels = "(GCF_900626175.2)")

# circos.track(ylim = c(0, 1), 
#     bg.col = NA, 
#     bg.border = "#00000044", 
#     track.height = 0.1)

circos.trackHist( 
  sectors = as.factor(genes$seqid),
  x = genes$start,
  #border = "#FFFFFF",
  bg.border = "#FFFFFF",
  bg.col = myBed$cols,
  #col = c("#FF0000", "#00FF00"),
  #col = "#C0C0C0",
  col = "#808080",
  border = NA,
  breaks = seq(1, 1e9, by = 2e6) )


circos.track(ylim = c(0, 1),
    bg.col = NA,
    bg.border = "#00000044",
    track.height = 0.15, bg.lty = 3, bg.lwd = 2)

circos.trackPoints(
  sectors = as.factor(rrna$seqid),
  x = rrna$start,
  y = rep(0.5, times = nrow(rrna)),
  #y = genes$start,
  #pch = "|",
  pch = 18,
  col = "#B2222211"
)

circos.trackPoints(
  sectors = as.factor(terps$seqid),
  x = terps$start,
  y = rep(0.2, times = nrow(terps)),
  #y = genes$start,
  #pch = "|",
  pch = 17,
  col = "#1E90FF11"
)

circos.track(ylim = c(0, 1),
    bg.col = NA,
    bg.border = "#00000044",
    track.height = 0.10, bg.lty = 3, bg.lwd = 1)

# circos.trackLines(
#   sectors = as.factor(terps$seqid),
#   x = terps$start,
#   y = terps$end,
#   #y = rep(0.2, times = nrow(terps)),
#   #y = genes$start,
#   #pch = "|",
# #  pch = 17,
#   col = "#1E90FF11"
# )


for( i in unique(terps$seqid) ){
  tmp <- terps[terps$seqid == i, ]

  circos.rect(
    xleft = tmp$start,
    ybottom = -0.0,
    xright = tmp$end,
    ytop = 0.5,
    sector.index = i,
    col = "#1E90FF11",
    border = "#1E90FF11",
    lwd = 4
  )
}

circos.link("NC_044374.1", 0,
            "NC_044379.1", c(1e3,4e7),
            col = "#1E90FF66", 
            #border = "black", 
            lwd = 2, lty = 2)

circos.link("NC_044374.1", 0,
            "NC_044375.1", c(1e3,6e7),
            col = "#FFD70066", 
            #border = "black", 
            lwd = 2, lty = 2)

circos.labels(c("NC_044378.1", "NC_044378.1"),
              x = c(30.1e6, 28.12e6), labels = c("A", "CBDAS"),
    side = "inside")

# circos.link("NC_044374.1", 0, "NC_044379.1", c(1e3,4e7),
#             col = "green", border = "black", lwd = 2, lty = 2)

#circos.text( x = 1, y = 200.5, labels = "Hi", sector.index = "NC_044371.1", track.index = 2)
#circos.text( x = 115e6, y = 200.5, labels = "GC", sector.index = "NC_044370.1", track.index = 2)
circos.text( x = 115e6, y = 0.5, labels = "Gene", sector.index = "NC_044370.1", track.index = 3)
circos.text( x = 115e6, y = 0.5, labels = "Line", sector.index = "NC_044370.1", track.index = 4)

#circos.yaxis( side = "right", sector.index =  "NC_044370.1", track.index = 2, labels.cex = 0.4)
circos.yaxis( side = "left", 
              at = seq(0, 300, by = 100),
              sector.index =  "NC_044371.1", track.index = 2, labels.cex = 0.4)

myLeg <- rep("", times = 10)
myLeg[c(2, 9)] <- c("High GC%", "Low GC%")

legend('bottomleft',
       #legend = c("T1", "", "", "T2"), 
       #legend = c("High GC", rep("", times = 8), "Low GC"),
       legend = myLeg,
       #fill = myBed$cols, 
       fill = viridisLite::magma(n=10, alpha = 0.4, begin = 1.0, end = 0.2),
       #fill = viridisLite::magma(n=10, alpha = 0.4, begin = 0.2, end = 1.0),
       border = NA,
       bty = "n", y.intersp = 0.5)

legend('bottomright',
       legend = c("rRNA", "Terpene"),
       col = c("#B22222", "#1E90FF"),
       #fill = c("#B22222", "#1E90FF"),
       #col = c("#B2222211", "#1E90FF11"),
       border = NA,
       pch = c(18, 17),
       pt.cex = 2,
       bty = "n",
       y.intersp = 1.6)


title( main = "Industrial hemp: CBDRx\n(GCF_900626175.2)", line = -1 )



circos.clear()

#dev.off()


```




## Read in data








```{r}
#cann <- read.table("../cann_motif_CBDRx_blastn_b2g.gff", sep = "\t")
cann <- read.table("./cann_motif_CBDRx_blastn_b2g.gff", sep = "\t")
colnames(cann)[c(1, 4, 5)] <- c("name", "start", "end")
cann[1:3, 1:8]
```



```{r}
nSeq <- 10
myBed <- data.frame( chrom = nucs$Id[1:nSeq], chromStart = 1, chromEnd = nucs$Length[1:nSeq])
apply(myBed , MARGIN = 2, class)
```







## Organize data


```{r}
df <- data.frame(name = nucs$Id, start = 1, end = nucs$Length)

#df <- df[-17,]
#sub("ref", "", df$name)
# df$name <- sub("ref\\|", "", df$name)
# df$name <- sub("\\|", "", df$name)
df$col <- "#B2222240"
df$col[4] <- "#9400D3"
df$col[9] <- "#40E0D040"

#df$gc <- (nucs$G[-17] + nucs$C[-17])/nucs$Length[-17]
df$gc <- (nucs$G + nucs$g + nucs$C + nucs$c)/nucs$Length
df$gc <- df$gc - min(df$gc)
df$gc <- df$gc/max(df$gc)
hist(df$gc)

df$col2 <- rgb(0, df$gc, 0, 0.4, names = NULL, maxColorValue = 1)

df[1:3, ]
```


## Plot graphic

```{r, fig.width=8, fig.height=8}

# png(filename = "myplot.png", width = 20, height = 20,
#    units = "cm", res = 300 )

#circos.par(start.degree = 0)
#circos.par(start.degree = 90)
#circos.par(start.degree = 180)

circos.genomicInitialize( data = df )

# circos.track(ylim = c(0, 1), 
#     bg.col = c("#FF000040", "#00FF0040", "#0000FF40"), 
#     bg.border = NA, track.height = 0.05)

# circos.track(ylim = c(0, 1), 
#     bg.col = rep(c("#B2222240", "#40E0D040", "#9400D3"), times = nrow(df)), 
#     bg.border = NA, track.height = 0.15)

circos.track(ylim = c(-10, 4), 
    bg.col = df$col2, 
    bg.border = NA, track.height = 0.15)

#my_chrom <- "NC_001147"
my_chrom <- df$name[2]
circos.points( 
  x = seq(1, df$end[df$name == my_chrom], length.out = 100), 
  y = seq(0, 1, length.out = 100), 
  sector.index = my_chrom,
  pch = 20, cex = 0.4
  )

my_chrom <- df$name[3]
circos.points( 
  x = seq(1, df$end[df$name == my_chrom], length.out = 100), 
  y = seq(0, 20, length.out = 100), 
  sector.index = my_chrom,
  pch = 20, cex = 0.4
  )


circos.track(ylim = c(0, 1), 
    bg.col = NA, 
    bg.border = "#00000044", track.height = 0.15)
    
circos.rect(xleft = cann$start, ybottom = 0.1,
            xright = cann$end, ytop = 0.8, 
            sector.index = cann$name[1])


# circos.barplot(value = seq(0, 1, length.out = 10),
#                pos = seq(1, 1e6, length.out = 10) - 0.5,
#                col = 1:10)

# circos.track(ylim = c(0, 255), 
#     bg.col = df$col2, 
#     bg.border = NA, track.height = 0.15)

text(x = 0, y = 0, "My genome")

#axis(side = 1)
#axis(side = 2)

circos.clear()

# dev.off()

```



```{r}
st99 <- Sys.time()
st99 - st1
```


